rust_src/src/buffers.rs

```diff
@@ -1149,6 +1171,41 @@ impl From<LispBufferOrCurrent> for LispBufferRef {
     }
 }
 
+/// Iterates over buffers by traversing over each `next` member of encountered
+/// buffers.
+pub struct LispBufferIter {
+    current: Option<LispBufferRef>,
+}
+
+impl LispBufferIter {
+    /// Start iteration at a given buffer.
+    #[allow(dead_code)]
+    pub fn with_start(buffer: LispBufferRef) -> Self {
+        Self {
+            current: Some(buffer)
+        }
+    }
+
+    /// Iterate over all buffers, i.e. start at the buffer defined by `all_buffers`.
+    pub fn all_buffers() -> Self {
+        unsafe {
+            Self {
+                current: LispBufferRef::from_ptr(all_buffers as *mut c_void),
```

Comment:
Note that the constructor iterating all buffers doesn't rely on the one for starting at a specified point, as I felt safer accounting for the possibility of `all_buffers` being null.